patches/tModLoader/Terraria/ModLoader/SystemLoader.cs

```diff
@@ -37,9 +40,31 @@ internal static void Unload()
 		SystemsByMod.Clear();
 	}
 
-	internal static void ResizeArrays()
+	internal static void ResizeArrays(bool unloading)
 	{
 		RebuildHooks();
+
+		if (unloading) {
+			return;
+		}
+		foreach (var mod in ModLoader.Mods) {
+			var reinitializeTypes = AssemblyManager.GetLoadableTypes(mod.Code)
+				.Where(t => t.GetAttribute<ReinitializeDuringResizeArraysAttribute>() != null)
+				//.Where(t => !t.IsAbstract && !t.ContainsGenericParameters) // static is IsAbstract and IsSealed
+				//.Where(t => t.IsAssignableTo(typeof(ILoadable)))
+				//.Where(t => t.GetConstructor(BindingFlags.Instance | BindingFlags.Public | BindingFlags.NonPublic, Type.EmptyTypes) != null) // has default constructor
+				.Where(t => AutoloadAttribute.GetValue(t).NeedsAutoloading)
```

Comment:
This one is tricky. We want it to work with `Autoload(false)` but not with `Autoload(Side = ...)` on the wrong side. I think it's best to remove this to reduce footguns. 

The `Side =` and `ModOrganizer.LoadSide` checks should really be applied to `GetLoadableTypes` but no-one has complained yet...